lib/widgets/compose_box.dart

```diff
@@ -128,14 +134,78 @@ class ComposeContentController extends ComposeController<ContentValidationError>
       : TextRange.collapsed(text.length);
   }
 
+  /// Inserts [newText] in [text], setting off with an empty line before and after.
+  ///
+  /// Assumes [newText] consists entirely of complete lines
+  /// (each line ends with a newline).
+  ///
+  /// Inserts at [_insertionIndex]. If that's zero, no empty line is added before.
+  ///
+  /// If there is already an empty line before or after, does not add another.
+  void _insertPadded(String newText) {
+    final i = _insertionIndex();
+    final textBefore = text.substring(0, i.start);
+    final String paddingBefore;
+    if (textBefore.isEmpty || textBefore == '\n' || textBefore.endsWith('\n\n')) {
+      paddingBefore = ''; // At start of input, or just after an empty line.
+    } else if (textBefore.endsWith('\n')) {
+      paddingBefore = '\n'; // After a complete but non-empty line.
+    } else {
+      paddingBefore = '\n\n'; // After an incomplete line.
+    }
+    final paddingAfter = text.substring(i.start).startsWith('\n') ? '' : '\n';
+    value = value.replaced(i, paddingBefore + newText + paddingAfter);
```

Comment:
This logic definitely calls for some unit tests.

I think this method would be fine to make public, and then it can be tested directly.